keps/sig-node/4815-dra-partitionable-devices/README.md

```diff
@@ -0,0 +1,2140 @@
+|          2g.10gb        |          2g.10gb        |          2g.10gb        |            X            |
++------------+------------+------------+------------+------------+------------+------------+------------+
+|          1g.10gb        |          1g.10gb        |          1g.10gb        |          1g.10gb        |
++------------+------------+------------+------------+------------+------------+------------+------------+
+|   1g.5gb   |   1g.5gb   |   1g.5gb   |   1g.5gb   |   1g.5gb   |   1g.5gb   |   1g.5gb   |     X      |
++------------+------------+------------+------------+------------+------------+------------+------------+
+| 1g.5gb+me  | 1g.5gb+me  | 1g.5gb+me  | 1g.5gb+me  | 1g.5gb+me  | 1g.5gb+me  | 1g.5gb+me  |     X      |
++------------+------------+------------+------------+------------+------------+------------+------------+
+```
+
+The most important thing to note from all of this is that partitioning a GPU
+into a set of MIG devices is not as simple as slicing it up into a set of
+fixed-size, non-overlapping partitions. Instead, one can think of a GPU as a
+"bag" of resources that can be pulled together in a discrete set of ways to
+construct a MIG device with a particular profile. When instatiating a MIG
+device with a particular profile, the set of resources available to other MIG
+devices (or the full GPU itself) gets depleted.
+
+In other words, the allocation of one MIG device may nullify the
+ability to allocate some subset of other possible MIG devices (but not all of
+them). A MIG device is considered allocatable so long as the full GPU has
+enough *unallocated* capacity to satisfy its capacity constraints across all
+dimensions. If any single dimension is unavailable, that MIG device cannot be
+allocated (but others might still be able to).
+
+For example, the following `ResourceClaim` can be used to select a set of
+non-overlapping MIG devices from a specific GPU.
+```yaml
+apiVersion: resource.k8s.io/v1alpha3
+kind: ResourceClaim
+metadata:
+  name: mig-devices
+spec:
+  spec:
+    devices:
+      requests:
+      - name: mig-1g-5gb-0
+        deviceClassName: mig.nvidia.com
+        selectors:
+        - cel:
+            expression: "device.attributes['gpu.nvidia.com'].profile == '1g.5gb'"
+      - name: mig-1g-5gb-1
+        deviceClassName: mig.nvidia.com
+        selectors:
+        - cel:
+            expression: "device.attributes['gpu.nvidia.com'].profile == '1g.5gb'"
+      - name: mig-2g-10gb
+        deviceClassName: mig.nvidia.com
+        selectors:
+        - cel:
+            expression: "device.attributes['gpu.nvidia.com'].profile == '2g.10gb'"
+      - name: mig-3g-20gb
+        deviceClassName: mig.nvidia.com
+        selectors:
+        - cel:
+            expression: "device.attributes['gpu.nvidia.com'].profile == '3g.20gb'"
+      constraints:
+      - requests: []
+        matchAttribute: "gpu.nvidia.com/parentUUID"
+```
+
+This would result in the following set of non-overlapping partitions:
```

Comment:
The scheduler might attempt to do this, but in doing so it would not be able to allocate all requests. I would then need to backtrack and try a different combination. Specific logic for doing this with MIG devices can be seen here: https://github.com/NVIDIA/mig-parted/blob/main/pkg/mig/config/config.go#L125. We would need to generalize this for the abstractions being introduced here.